src/typechecker.erl

```diff
@@ -4734,46 +4737,112 @@ get_record_info_type({call, Anno, {atom, _, record_info},
-                                end,
-                                erlang:raise(error, Error, Trace)
-                        end;
-              (_Function, Errors) ->
-                  Errors
-          end, [], ParseData#parsedata.functions),
+    AllErrors = lists:foldr(fun (Function, Errors) ->
+                                    type_check_form_with_timeout(Function, Errors, StopOnFirstError, Env, Opts)
+                            end, [], ParseData#parsedata.functions),
     lists:reverse(AllErrors).
 
+
+%% @doc `type_check_form_with_timeout' is a workaround meant to improve usability in the
+%% presence of possible infinite loops or bugs in the typechecker.
+%%
+%% When Gradualizer is run interactively or in the foreground,
+%% detection of an infinite loop is unavoidable by the user seeing that the tool is stuck.
+%%
+%% However, when Gradualizer is run in the background (as an integration with editor / IDE)
+%% we might not realise it's stuck in an infinite loop at all or do so only after the machine's fan
+%% starts up. Moreover, in this case we'll not receive feedback from the tool.
+%%
+%% To avoid the user experience problem, it's better to opportunistically try performing the check,
+%% but in the light of it taking too long, just forcibly break the infinite loop and report
+%% a Gradualizer (NOT the checked program!) error.
+type_check_form_with_timeout(Function, Errors, StopOnFirstError, Env, Opts) ->
+    %% TODO: make FormCheckTimeOut configurable
+    FormCheckTimeOut = ?form_check_timeout_ms,
+    ?verbose(Env, "Spawning async task...~n", []),
+    TaskF = fun () ->
+                    type_check_form(Function, Errors, StopOnFirstError,
+                                    Env, Opts)
+            end,
+    %% We could tweak this to log and proceed with the next form instead of crashing...
+    DownF = fun (Down) ->
+                    {error_trace, Down, []}
+            end,
+    case timeout(TaskF, FormCheckTimeOut, DownF) of
+        timeout ->
+            ?verbose(Env, "Form check timeout on ~s~n",
+                     [gradualizer_fmt:form_info(Function)]),
+            [{internal_error, form_check_timeout, Function} | Errors];
+        {crash, Crash, St} ->
+            ?verbose(Env, "Task reported crash on ~s~n",
+                     [gradualizer_fmt:form_info(Function)]),
+            io:format("Crashing...~n"),
+            erlang:raise(throw, Crash, St);
+        {error_trace, Error, Trace} ->
+            ?verbose(Env, "Task reported error with trace from ~s~n",
+                     [gradualizer_fmt:form_info(Function)]),
+            erlang:raise(error, Error, Trace);
+        {errors, Errors1} ->
+            ?verbose(Env, "Task returned from ~s with ~p~n",
+                     [gradualizer_fmt:form_info(Function), Errors1]),
+            Errors1
+    end.
+
+-spec timeout(fun(() -> any()), non_neg_integer(), fun(({'DOWN', _, _, _, _}) -> any())) -> any().
+timeout(TaskF, Timeout, DownF) ->
```

Comment:
This abstraction is smaller than the full Elixir `task` :+1: but it's still an abstraction only used once. I'm not a fan of that. (Small functions are good, but wrapping code in extra funs only used once is mainly added complexity.)

I'd rather put the spawn_monitor in type_check_form_with_timeout and replace the case..of with this receive here. The down fun isn't necessary.